subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java

```diff
@@ -0,0 +1,233 @@
+ */
+package org.gradle.plugins.ide.internal.resolver;
+
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Table;
+import org.gradle.api.Action;
+import org.gradle.api.artifacts.ArtifactCollection;
+import org.gradle.api.artifacts.ArtifactView;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.component.ComponentArtifactIdentifier;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ComponentSelector;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.artifacts.dsl.DependencyHandler;
+import org.gradle.api.artifacts.result.ArtifactResolutionResult;
+import org.gradle.api.artifacts.result.ArtifactResult;
+import org.gradle.api.artifacts.result.ComponentArtifactsResult;
+import org.gradle.api.artifacts.result.ResolvedArtifactResult;
+import org.gradle.api.artifacts.result.UnresolvedDependencyResult;
+import org.gradle.api.component.Artifact;
+import org.gradle.api.specs.Spec;
+import org.gradle.api.specs.Specs;
+import org.gradle.jvm.JvmLibrary;
+import org.gradle.language.base.artifact.SourcesArtifact;
+import org.gradle.language.java.artifact.JavadocArtifact;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Adapts Gradle's dependency resolution engine to the special needs of the IDE plugins.
+ * Allows adding and subtracting {@link Configuration}s, working in offline mode and downloading sources/javadoc.
+ */
+public class IdeDependencySet {
+    private final DependencyHandler dependencyHandler;
+    private final Collection<Configuration> plusConfigurations;
+    private final Collection<Configuration> minusConfigurations;
+
+    public IdeDependencySet(DependencyHandler dependencyHandler, Collection<Configuration> plusConfigurations, Collection<Configuration> minusConfigurations) {
+        this.dependencyHandler = dependencyHandler;
+        this.plusConfigurations = plusConfigurations;
+        this.minusConfigurations = minusConfigurations;
+    }
+
+    public void visit(IdeDependencyVisitor visitor) {
+        if (plusConfigurations.isEmpty()) {
+            return;
+        }
+        new IdeDependencyResult().visit(visitor);
+    }
+
+    /*
+     * Tries to minimize the number of requests to the resolution engine by batching up requests
+     * for sources/javadoc.
+     *
+     * There is still some inefficiency because the ArtifactCollection interface does not provide
+     * detailed failure results, so we have to fall back to the more expensive ResolutionResult API.
+     * We should fix this, as other IDE vendors will face the same problem.
```

Comment:
Is this PR the sort of thing we'd need? https://github.com/gradle/gradle/pull/2736

Reply:
Yes, I was thinking about picking that back up to support the IDEA and Android Studio teams as well as our own code.